boa/src/jsstring.rs

```diff
@@ -0,0 +1,636 @@
+    utf16!("find"),
+    utf16!("findIndex"),
+    utf16!("flat"),
+    utf16!("flatMap"),
+    utf16!("forEach"),
+    utf16!("includes"),
+    utf16!("indexOf"),
+    utf16!("join"),
+    utf16!("map"),
+    utf16!("reduce"),
+    utf16!("reduceRight"),
+    utf16!("reverse"),
+    utf16!("shift"),
+    utf16!("slice"),
+    utf16!("some"),
+    utf16!("sort"),
+    utf16!("unshift"),
+    utf16!("push"),
+    utf16!("pop"),
+    // String object
+    utf16!("String"),
+    utf16!("charAt"),
+    utf16!("charCodeAt"),
+    utf16!("concat"),
+    utf16!("endsWith"),
+    utf16!("lastIndexOf"),
+    utf16!("match"),
+    utf16!("matchAll"),
+    utf16!("normalize"),
+    utf16!("padEnd"),
+    utf16!("padStart"),
+    utf16!("repeat"),
+    utf16!("replace"),
+    utf16!("replaceAll"),
+    utf16!("search"),
+    utf16!("split"),
+    utf16!("startsWith"),
+    utf16!("substring"),
+    utf16!("toLowerString"),
+    utf16!("toUpperString"),
+    utf16!("trim"),
+    utf16!("trimEnd"),
+    utf16!("trimStart"),
+    // Number object
+    utf16!("Number"),
+    // Boolean object
+    utf16!("Boolean"),
+    // RegExp object
+    utf16!("RegExp"),
+    utf16!("exec"),
+    utf16!("test"),
+    utf16!("flags"),
+    utf16!("index"),
+    utf16!("lastIndex"),
+    // Symbol object
+    utf16!("Symbol"),
+    utf16!("for"),
+    utf16!("keyFor"),
+    utf16!("description"),
+    utf16!("[Symbol.toPrimitive]"),
+    // Map object
+    utf16!("Map"),
+    utf16!("clear"),
+    utf16!("delete"),
+    utf16!("has"),
+    utf16!("size"),
+    // Set object
+    utf16!("Set"),
+    // Reflect object
+    utf16!("Reflect"),
+    // Error objects
+    utf16!("Error"),
+    utf16!("TypeError"),
+    utf16!("RangeError"),
+    utf16!("SyntaxError"),
+    utf16!("ReferenceError"),
+    utf16!("EvalError"),
+    utf16!("URIError"),
+    utf16!("message"),
+    // Date object
+    utf16!("Date"),
+    utf16!("toJSON"),
+];
+
+const MAX_CONSTANT_STRING_LENGTH: usize = {
+    let mut max = 0;
+    let mut i = 0;
+    while i < CONSTANTS_ARRAY.len() {
+        let len = CONSTANTS_ARRAY[i].len();
+        if len > max {
+            max = len;
+        }
+        i += 1;
+    }
+    max
+};
+
+thread_local! {
+    static CONSTANTS: FxHashSet<JsString> = {
+        let mut constants = FxHashSet::default();
+
+        for s in CONSTANTS_ARRAY.iter() {
+            let s = JsString::from_slice_skip_interning(s);
+            constants.insert(s);
+        }
+
+        constants
+    };
+}
+
+/// The inner representation of a [`JsString`].
+#[repr(C)]
+struct Inner {
+    /// The utf16 length.
+    len: usize,
+
+    /// The number of references to the string.
+    ///
+    /// When this reaches `0` the string is deallocated.
+    refcount: Cell<usize>,
+
+    /// An empty array which is used to get the offset of string data.
+    data: [u16; 0],
+}
+
+#[derive(Finalize)]
+pub struct JsString {
+    ptr: NonNull<Inner>,
+    phantom: PhantomData<Inner>,
```

Comment:
@jedel1043 what's PhantomData here and why is it used? First time I've come across it

Reply:
It's a hint to tell the compiler that dropping a `JsString` may drop `Inner`. However, I was just copying the implementation of `Rc` from the std, and after some research I realized it may not be necessary in this case, it's just needed when you have self-referential structures or lifetimes that aren't used inside the struct definition. (Reference: https://stackoverflow.com/a/58034644)
I'll still try to test if removing the `PhantomData` works without problems.